### PR TITLE
couchdb version upgrade preparedness

### PR DESCRIPTION
**Pull request by shyamkumarlchauhan**
https://dimagi-dev.atlassian.net/browse/IIO-925
Couchdb version upgrade readiness
just make changes in in public.yml of requirement environment for below value

couch_version: 3.1.1

##### ENVIRONMENTS AFFECTED
None. Code changes to prepare for couchdb upgrade on required environments

### CODE REVIEW

**Review by snopoke — COMMENTED** [review]
This is almost an exact copy of the couchdb2 role. Apart from renaming couchdb2 to couchdb3 the only changes seem to be:

- 3 tasks in main.yml to update /etc/init.d/couchdb3
- removing the couchdb2-restarter.sh.j2 file

If those are the only changes necessary then we should not have a separate role.

**Comment by snopoke** [comment]
@shyamkumarlchauhan I see this is still WIP so probably ignore my last comment.

File: docs/services/couchdb/upgrade.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    ```
+
+3. Upgrade the CouchDB on rolling basis one by one on each couchDB nodes
+
+    ```
+    $ cchq <env> ap deploy_couchdb2.yml --limit=couch0(N)
+    ```
+
+4. Upgrade the  one node at a time and reboot the node before applying the upgrade on another couchDB node.
+   Perform the after-reboot once the node has been rebooted.
```
[the remark this answers is not in the document]
Comment (shyamkumarlchauhan) [reply docs/services/couchdb/upgrade.md:28]: Reboot is required as restart of only couchdb service is not enough . as epmd process is occupying the older PID of couchdb process

Below is the message from log file :

Nov 11 06:01:56 couchtest1 ansible[16231]: root : /etc/init.d/couchdb2 start
Nov 11 06:01:57 couchtest1 epmd[7981]: epmd: node name already occupied couchdb
Nov 11 06:01:58 couchtest1 /etc/init.d/couchdb[16324]: couchdb2 is not running
Nov 11 06:01:58 couchtest1 ansible[16330]: root : /etc/init.d/couchdb2 status

**Comment by sravfeyn** [comment]
One change in version 3 I noticed was that the admin-party feature is killed, so we would have to provide the admin user/password in default config. Is this already the case in our previous couchdb2 code?

**Comment by sravfeyn** [comment]
Don't we need to update https://github.com/dimagi/couchdb-cluster-admin ?

File: docs/services/couchdb/upgrade.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (shyamkumarlchauhan) [reply docs/services/couchdb/upgrade.md:28]: > One change in version 3 I noticed was that the admin-party feature is killed, so we would have to provide the admin user/password in default config. Is this already the case in our previous couchdb2 code?

Yes we have been previosly supplying the user and password in couchdb2.

 user: "{{ couchdb2.username }}"
 password: "{{ couchdb2.password }}"

**Comment by shyamkumarlchauhan** [comment]
> Don't we need to update https://github.com/dimagi/couchdb-cluster-admin ?

i believe we use external role for this https://github.com/andrewrothstein/ansible-couchdb-cluster

**Comment by shyamkumarlchauhan** [comment]
> One change in version 3 I noticed was that the admin-party feature is killed, so we would have to provide the admin user/password in default config. Is this already the case in our previous couchdb2 code?

Yes we have been previosly supplying the user and password in couchdb2.

user: "{{ couchdb2.username }}"
password: "{{ couchdb2.password }}"

File: src/commcare_cloud/ansible/roles/couchdb2/tasks/main.yml [hunk 2]
```diff
@@ -39,11 +42,20 @@
...
+  when: couch_version == '2.3.1' and inventory_hostname == groups.couchdb2.0 and item != groups.couchdb2.0
   with_items: "{{ groups.couchdb2 }}"
   tags:
     - add_couch_nodes
 
+- name: Add nodes in couchdb3 cluster
+  shell: "curl -X PUT http{% if couchdb_secure %}s{% endif %}://{{ couchdb2.username }}:{{ couchdb2.password }}@{{ inventory_hostname|ipaddr }}:{{ couchdb_port }}/_node/_local/_nodes/couchdb@{{ item|ipaddr }} -d {}"
+  args:
+    warn: false
+  when: couch_version == '3.1.1' and inventory_hostname == groups.couchdb2.0 and item != groups.couchdb2.0
```
Comment (snopoke) [inline src/commcare_cloud/ansible/roles/couchdb2/tasks/main.yml:54]:
```suggestion
  when: couch_version is version('3.0.0', '>=') and inventory_hostname == groups.couchdb2.0 and item != groups.couchdb2.0
```

File: src/commcare_cloud/ansible/roles/couchdb2/tasks/main.yml [hunk 3]
```diff
@@ -39,11 +42,20 @@
     force_basic_auth: true
     status_code: 201, 409 # 409 means already set up (conflict)
     body: '{}'
-  when: inventory_hostname == groups.couchdb2.0 and item != groups.couchdb2.0
+  when: couch_version == '2.3.1' and inventory_hostname == groups.couchdb2.0 and item != groups.couchdb2.0
```
Comment (snopoke) [inline src/commcare_cloud/ansible/roles/couchdb2/tasks/main.yml:45]:
```suggestion
  when: couch_version is version('3.0.0', '<') and inventory_hostname == groups.couchdb2.0 and item != groups.couchdb2.0
```

File: src/commcare_cloud/ansible/roles/couchdb2/tasks/main.yml [hunk 4]
```diff
@@ -39,11 +42,20 @@
...
     - add_couch_nodes
 
+- name: Add nodes in couchdb3 cluster
+  shell: "curl -X PUT http{% if couchdb_secure %}s{% endif %}://{{ couchdb2.username }}:{{ couchdb2.password }}@{{ inventory_hostname|ipaddr }}:{{ couchdb_port }}/_node/_local/_nodes/couchdb@{{ item|ipaddr }} -d {}"
+  args:
+    warn: false
+  when: couch_version == '3.1.1' and inventory_hostname == groups.couchdb2.0 and item != groups.couchdb2.0
+  with_items: "{{ groups.couchdb2 }}"
+  tags:
+    - add_couchdb3_nodes
```
Comment (snopoke) [inline src/commcare_cloud/ansible/roles/couchdb2/tasks/main.yml:57]: no need to change the tag
```suggestion
    - add_couch_nodes
```

**Review by snopoke — COMMENTED** [review]
I recently tried to install couch (2.3.1) on a new VM with Ubuntu 18.04 and had a few issues:
- unsupported version of erlang
- unable to install "couch-libmozjs185-dev" requirement

Have you come across any of those issues?

**Pull request closed without merging**